addons/controllers/clusterbootstrap_controller.go

```diff
@@ -1464,3 +1539,157 @@ func (r *ClusterBootstrapReconciler) reconcileClusterProxyAndNetworkSettings(clu
 
 	return nil
 }
+
+func (r *ClusterBootstrapReconciler) clusterIsReadyForDeletion(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (bool, error) {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), clusterBootstrap)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "failed to lookup clusterbootstrap")
+		return false, err
+	}
+	if apierrors.IsNotFound(err) {
+		log.Info("cluster bootstrap not found")
+	} else {
+		remoteClient, err := util.GetClusterClient(r.context, r.Client, r.Scheme, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			return false, err
+		}
+		if hasPackageInstalls(r.context, remoteClient, cluster, r.Config.SystemNamespace, clusterBootstrap.Spec.AdditionalPackages, log) {
+			err = r.removeAdditionalPackageInstalls(remoteClient, cluster, clusterBootstrap, log)
+			if err != nil {
+				return false, err
+			}
+			return false, nil
+		}
+	}
+	log.Info("cluster is ready for deletion")
+	return true, nil
+}
+
+func (r *ClusterBootstrapReconciler) removeFinalizersFromClusterResources(cluster *clusterapiv1beta1.Cluster, log logr.Logger) error {
+	log.Info("removing finalizer from clusterbootstrap if present")
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), clusterBootstrap)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "failed to lookup clusterbootstrap")
+		return err
+	}
+	if !apierrors.IsNotFound(err) {
+		err = r.removeFinalizer(clusterBootstrap)
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Info("removing finalizer from cluster kubeconfig secret if present")
+	clusterKubeConfigSecret := &corev1.Secret{} // TODO (adduarte): test without deleting clusterKubeConfig
+	key := client.ObjectKey{Namespace: cluster.Namespace, Name: secretutil.Name(cluster.Name, secretutil.Kubeconfig)}
+	err = r.Client.Get(r.context, key, clusterKubeConfigSecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "unable to fetch cluster kubeconfig secret")
+		return err
+	}
+	if !apierrors.IsNotFound(err) {
```

Comment:
Check if err != nil or move into block above

Reply:
will refactor to make it easier to read. 
Basically at this point if we have no errors (err=nil) we remove the finalizer, but if have err=IsNotFound we do nothing, because there is no kubeconfig, which means it was already deleted, somehwere else. 
so we can proceed with removing the rest of the finalizers.

Reply:
done
refactored code as suggested